### ISSUE TITLE
How to not include a value in a segment

### ISSUE THREAD

**Issue by RJSchmertz** [indice-co/EDI.Net#108]
I currently have an X12 834 writer setup.  
I have the following class that properly outputs `INS*Y*18*021*28*A*C**RT****~`

I want to use the same writer/classes for another client, but they don't want the last 4 values in that segment and their ingest system fails if I send them the above value when they aren't expecting those last values.

Is there a  way to have the same classes (with additional attributes or what ever it takes) to output:
`INS*Y*18*021*28*A*C**RT~`?

```csharp
[EdiSegment, EdiSegmentGroup("INS", SequenceEnd = "LS")]
public class MemberDetailLoop //2000 (10,000)
{
    [EdiValue("X(1)", Path = "INS/0", Description = "INS01 - Yes/No Condition Response Code (Y = Subscriber N = Dependents = Members)")]
    public string ResponseCode { get; set; }

    [EdiValue("X(2)", Path = "INS/1", Description = "INS02 - Individual Relationship Code")]
    public string IndividualRelationshipCode { get; set; }

    [EdiValue("X(2)", Path = "INS/2", Description = "INS03 - Maintenance Type Code")]
    public string MaintenanceTypeCode { get; set; }

    [EdiValue("X(2)", Path = "INS/3", Description = "INS04 - Maintenance Reason Code")]
    public string MaintenanceReasonCode { get; set; }

    [EdiValue("X(1)", Path = "INS/4", Description = "INS05 - Benefit Status Code")]
    public string BenefitStatusCode { get; set; }

    [EdiValue("X(1)", Path = "INS/5", Description = "INS06 - Medicare Status Code")]
    public string MedicareStatusCode { get; set; }

    [EdiValue("X(2)", Path = "INS/6", Description = "INS07 - Consolidated Omnibus Budget Reconciliation Act (COBRA) Qualifyinge")]
    public string CobraQualifyingeCode { get; set; }

    [EdiValue("X(2)", Path = "INS/7", Description = "INS08 - Employment Status Code")]
    public string EmploymentStatusCode { get; set; }

    [EdiValue("X(2)", Path = "INS/8", Description = "INS09 - Student Status Code (This section is significant only if INS02 = 19 and if the child's age falls between the benefit's overage dependent age range.)")]
    public string StudentStatusCode { get; set; }

    [EdiValue("X(2)", Path = "INS/9", Description = "INS10 - Handicapped Indicator")]
    public string HandicappedIndicator { get; set; }

    [EdiValue("X(3)", Path = "INS/10", Description = "INS11 - Date Time Period Format Qualifier")]
    public string DateTimePeriodFormatQualifier { get; set; }

    [EdiValue("9(8)", Path = "INS/11", Format = "yyyyMMdd", Description = "INS12 - Date Time Period")]
    public DateTime? DateTimePeriod { get; set; }
}
```

I was trying to figure out how `EdiCondition()` works and if I could use that, but I couldn't figure out how to get it to do this.  I could base the logic off of the value in "ISA/7" if that helps, it is the ReceiverID in an upper scope InterchangeControlLoop

```csharp
        [EdiValue("X(15)", Path = "ISA/7", Description = "ISA08 - Premier Access Federal Tax ID Number to be used as an identifierD")]
        public string ReceiverId { get; set; }
```

So maybe if ISA/7 is 1234, include these values, if ISA/7 is 5678, don't include.
Or maybe there is an easier way.

Thank you

**Comment by cleftheris** [comment]
@RJSchmertz Hi and thanks for your interest in EDI.Net

Can you please tell me which version of the library you are using? 

By the way the condition attribute only works for container classes (for Elements, Segments, groups of Segments etc.)

**Comment by RJSchmertz** [comment]
Looks like we are on 1.3.1 right now. I'd be willing to try an upgrade if it helps.

I'm pretty new to building these, and inherited the current project.  Is there any container smaller than a segment (one line in the output file)?  Could I accomplish this by having multiple containers in the same segment?

**Comment by cleftheris** [comment]
@RJSchmertz you see what you are looking for is support for "compression". This is a new feature that comes out of the box in `v1.7.0` and is on by default.

You can turn it off by using the flag on the Serializer like so:

```csharp
new EdiSerializer() { EnableCompression = false }.Serialize(writer, interchange);
```

Hope this helps.

**Comment by RJSchmertz** [comment]
Will this only remove trailing values that are empty, or if there is an empty value in the middle it will remove it as well?
Given the first example `INS*Y*18*021*28*A*C**RT****~`
Will it only remove the 4 trailing values, or would the empty value before RT also be removed?

I really appreciate the quick responses. I will look at updating to 1.7.0 now.

**Comment by cleftheris** [comment]
Compression is implemented as the industry standard and omits any trailing empty elements not middle ones.it should only omit the trailing values in your case. If not please report back as it would be a bug :-)

**Comment by RJSchmertz** [comment]
This is definitely fixing most of my problems and is working as you describe. Thank you a ton!

I'm still trying to play with `EdiCondition` for removing entire segments.
I have a MemberDetailLoop that one of it's properties is a LanguageSegment.
I'd like to conditionally not include the segment

This is kind of what I've been trying (applying attribute to either the Segment Property on the parent, or the Segment object itself), but it seems to move it around the output rather than removing it. `ISA/7` is ReceiverId, so was seeing if I could do logic based on who the receiver is.
I think I am using it incorrectly.

```csharp
[EdiSegment, EdiSegmentGroup("INS", SequenceEnd = "LS")]
public class MemberDetailLoop //2000 (10,000)
{
	//... some other values

	//[EdiCondition("04-1045815     ", Path = "ISA/7", CheckFor = EdiConditionCheckType.NotEqual)]
    // Remove for receiver 04-1045815
	public LanguageSegment Langugage { get; set; } 
}

[EdiSegment, EdiPath("LUI")]
// Remove for receiver 04-1045815
//[EdiCondition("04-1045815     ", Path = "ISA/7", CheckFor = EdiConditionCheckType.NotEqual)]
public class LanguageSegment
{
    [EdiValue("X(2)", Path = "LUI/0", Description = "LUI01 - Identification Code Qualifier")]
    public string LanguageIdentificationCodeQualifier { get; set; }

    [EdiValue("X(80)", Path = "LUI/1", Description = "LUI02 - Identification Code")]
    public string LanguageIdentificationCode { get; set; }

    [EdiValue("X(1)", Path = "LUI/2", Description = "LUI03 - Language Description")]
    public string LanguageDescription { get; set; }

    [EdiValue("9(2)", Path = "LUI/3", Description = "LUI04 - Use of Language Indicator")]
    public string UseOfLanguageIndicator { get; set; }
}
```

With compression, I've been able to remove it if I set the segment object to null before serializing, but was seeing if I could do it via attributes.

Thank you again

**Comment by cleftheris** [comment]
Conditional attribute is designed to work with de-serialization. You can create logic inside your properties and conditionally return null using the getters and a little c#.

What you are trying is hard to achieve with a single multipurpose model. Try multiple models for each interchange type.

I am going to close this issue. But, If you got additional questions please open a new issue for each because it’s easier for others to find them.

**Comment by RJSchmertz** [comment]
You were very helpful. Thank you so much. Is there a donate button somewhere?

**Comment by cleftheris** [comment]
No I don't have something like that yet, but thanks for asking.

**Comment by michaelachrisco** [comment]
Hello @RJSchmertz, At one point I was working on the 834 and got the library to create the 834 via: https://github.com/michaelachrisco/EDI.Net/blob/master/test/indice.Edi.Tests/Models/X12_834.cs 

I was able to remove the trailing values in an older version of the library by:
```C#
            // serialize to file.
            using (MemoryStream stream = new MemoryStream())
            {
                using (StreamWriter textWriter = new StreamWriter(stream))
                {
                    stream.Position = 0;
                    using (EdiTextWriter ediWriter = new EdiTextWriter(textWriter, grammar))
                    {
                        ediSerializer.Serialize(ediWriter, _interchange);
                    }
                    string UnformattedEDI = System.Text.Encoding.Default.GetString(stream.ToArray());
                    //\b\*+\~
                    myString = rgx.Replace(UnformattedEDI, "~");
                }
            }

```

I don't recommend it if you can just upgrade but it does work..

**Comment by RJSchmertz** [comment]
I was able to upgrade and was able to clear it with my other clients to accept the trimmed segments. (Thank you cleftheris for supplying supplemental information that this was pretty standard in the industry, that enabled my conversations even more).
I love seeing creative ideas on solving problems and solving those, thank you @michaelachrisco .
I was going to ask my employer if they were comfortable with me sharing back the 834 format we've ended up with as well.  I'll compare to yours as well for improvements.

Thanks again guys :)
